File: mo-common/mo-common-core/src/main/java/ink/ckx/mo/common/core/result/Result.java
package ink.ckx.mo.common.core.result;

import lombok.Data;
import lombok.NoArgsConstructor;

import java.io.Serializable;
import java.util.Optional;

/**
 * @author chenkaixin
 * @description
 * @since 2023/10/19
 */
@NoArgsConstructor
@Data
public class Result<T> implements Serializable {

    private String code;

    private String msg;

    private T data;

    public Result(String code, String msg) {
        this(code, msg, null);
    }

    public Result(String code, String msg, T data) {
        this.code = code;
        this.msg = msg;
        this.data = data;
    }

    public static <T> Result<T> success() {
        return success(null);
    }

    public static <T> Result<T> success(T data) {
        return new Result<>(ResultCode.SUCCESS.getCode(), ResultCode.SUCCESS.getMsg(), data);
    }

    public static <T> Result<T> fail() {
        return Result.fail(ResultCode.FAIL);
    }

    public static <T> Result<T> fail(String msg) {
        return new Result<>(ResultCode.FAIL.getCode(), msg);
    }

    public static <T> Result<T> fail(ResultCode resultCode) {
        return new Result<>(resultCode.getCode(), resultCode.getMsg());
    }

    public static <T> Result<T> fail(ResultCode resultCode, String msg) {
        return new Result<>(resultCode.getCode(), Optional.ofNullable(msg).orElse(resultCode.getMsg()));
    }

    public static boolean isSuccess(Result<?> result) {
        return result != null && ResultCode.SUCCESS.getCode().equals(result.getCode());
    }
}

File: mo-common/mo-common-core/src/main/kotlin/ink/ckx/mo/common/core/result/Result.kt
package ink.ckx.mo.common.core.result

import java.io.Serializable
import java.util.*

/**
 * @author chenkaixin
 * @description
 * @since 2023/10/19
 */
class Result<T>(
    val code: String? = null,
    val msg: String? = null,
    val data: T? = null,
) : Serializable {

    companion object {
        fun <T> success(): Result<T?> {
            return success(null)
        }

        fun <T> success(data: T): Result<T> {
            return Result(ResultCode.SUCCESS.code, ResultCode.SUCCESS.msg, data)
        }

        fun <T> fail(): Result<T> {
            return fail(ResultCode.FAIL)
        }

        fun <T> fail(msg: String): Result<T> {
            return Result(ResultCode.FAIL.code, msg)
        }

        fun <T> fail(resultCode: ResultCode): Result<T> {
            return Result(resultCode.code, resultCode.msg)
        }

        fun <T> fail(resultCode: ResultCode, msg: String?): Result<T> {
            return Result(resultCode.code, Optional.ofNullable(msg).orElse(resultCode.msg))
        }

        fun isSuccess(result: Result<*>): Boolean {
            return ResultCode.SUCCESS.code == result.code
        }
    }
}